Implement `Location`'s custom internal methods

yvt opened the pull request:
This PR partly resurrects #16501 and introduces the use of principals object to associate objects and Realms with origins. Using this infrastructure, this PR implements [the custom internal methods][1] of the `Location` interface, which is "maybe-cross-origin".

Unimplemented/incomplete things:

 - Other maybe-cross-origin interfaces, namely `WindowProxy` and `DissimilarWindowLocation`, aren't implemented correctly yet (causing most test cases of `tests/wpt/web-platform-tests/html/browsers/origin/cross-origin-objects/cross-origin-objects.html` to fail). 
     - `WindowProxy`: #28556
 - [The "perform a security check" operation][2] and `Location`'s non-cross-origin properties' relevant `Document` origin checks aren't implemented either (not sure if they are covered by the existing tests).
 - There are a slight deviation from the standard and inefficiency in `CrossOriginGetOwnPropertyHelper`'s current implementation. 
     - #28557

[1]: https://html.spec.whatwg.org/multipage/#the-location-interface
[2]: https://html.spec.whatwg.org/multipage/browsers.html#integration-with-idl

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #16243 and make some progress in #2382

---
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

avadacatavra committed:
added JSPrincipal bindings

avadacatavra committed:
added xow security callbacks

avadacatavra committed:
added wrapper selection and subsumes logic

avadacatavra committed:
changed to use globalscope origin

avadacatavra committed:
investigating failures after filtering wrapper integration with mozjs

avadacatavra committed:
debugging

avadacatavra committed:
added dom exception callback stuff

avadacatavra committed:
added in cross origin type checks. assert_throws isn't working yet

avadacatavra committed:
frames override is the worst

avadacatavra committed:
show which tests work

avadacatavra committed:
xow tests reflect current state

avadacatavra committed:
debugging

avadacatavra committed:
enumerate should work. but is a dirty hack

yvt committed:
Merge remote-tracking branch 'upstream/master' into feat-cow-infra

`tests/wpt/web-platform-tests/html/browsers/origin/cross-origin-objects/cross-origin-objects.html`
was reverted to the upstream version.

yvt committed:
fix: accommodate to the modern age

yvt committed:
feat(script): add `CrossOrigin*able` attributes to `Window` and `Location`'s members

yvt committed:
Merge remote-tracking branch 'upstream/master' into feat-cow-infra

yvt committed:
chore(deps): update mozjs

 - 798c5b6: Bring `RustJSPrincipals` back

yvt committed:
feat(script): add `{DissimilarOriginWindow, PaintWorkletGlobalScope, TestWorkletGlobalScope}::origin`

yvt committed:
feat(script): remove the call to `JS_SetWrapObjectCallbacks`

We don't make CCWs anymore.

yvt committed:
fix(script): remove debug printing

yvt committed:
doc(script): prettify a comment

yvt committed:
fix(script): apply some of the changes requested in the review comments of #16501

yvt committed:
fix(script): implement the destroy-principals callback correctly

yvt committed:
feat(script): use `GlobalScope::origin` when creating a principals object

The concrete types of `[Global]` DOM interfaces have `origin` methods,
which were used before this commit. Some of them just delegate to
`GlobalScope::origin` while others are implemented differently. This
commit changes the created principals objects' associated origins in the
following way:

 - `DedicatedWorkerGlobalScope` - was `WorkerGlobalScope::worker_url`
 - `DissimilarOriginWindow` - no change
 - `PaintWorkletGlobalScope` - no change
 - `ServiceWorkerGlobalScope` - was `ServiceWorkerGlobalScope::scope_url`
 - `TestWorkletGlobalScope` - no change
 - `Window` - no change

yvt committed:
Revert "feat(script): add `{DissimilarOriginWindow, PaintWorkletGlobalScope, TestWorkletGlobalScope}::origin`"

This reverts commit 5959c2ef9b343edc5562f4efee2877334672e841.

yvt committed:
refactor(script): remove `{DedicatedWorkerGlobalScope, ServiceWorkerGlobalScope}::origin`

yvt committed:
refactor(script): move `crate::dom::bindings::{utils → principals)::ServoJSPrincipal`

yvt committed:
refactor(script): rename `ServoJSPrincipal` to `ServoJSPrincipals`

yvt committed:
refactor(script): auto ref-count `ServoJSPrincipals`

yvt committed:
Merge remote-tracking branch 'upstream/master' into feat-cow-infra

yvt committed:
feat(script): implement some of the non-ordinary internal methods of `Location`

<https://html.spec.whatwg.org/multipage/#the-location-interface>

 - `[[GetPrototypeOf]]`: not yet
 - `[[SetPrototypeOf]]`: not yet
 - `[[IsExtensible]]`: `proxyhandler::is_extensible`
 - `[[PreventExtensions]]`: `proxyhandler::prevent_extensions`
 - `[[GetOwnProperty]]`: `CGDOMJSProxyHandler_getOwnPropertyDescriptor` (updated)
 - `[[DefineOwnProperty]]`: `CGDOMJSProxyHandler_defineProperty` (updated)
 - `[[Get]]`: `CGDOMJSProxyHandler_get` (updated)
 - `[[Set]]`: not yet
 - `[[Delete]]`: `CGDOMJSProxyHandler_delete` (updated)
 - `[[OwnPropertyKeys]]`: `CGDOMJSProxyHandler_ownPropertyKeys` (updated)

yvt committed:
test: re-enable `/html/browsers/origin/cross-origin-objects/cross-origin-objects.html`

yvt committed:
feat(script): Implement `[[{Get,Set}PrototypeOf]]` for `Location`

yvt committed:
feat(script): enable `js::ProxyOptions::setLazyProto` for maybe-cross-origin objects

Setting the lazy proto option allows proxy handlers to provide dynamic
prototype objects. This is necessary for the customization of
`ProxyTraps::{get,set}PrototypeOf` to actually take effect.

yvt committed:
test: update expectations

No improvements are seen in `/html/browsers/origin/cross-origin-
objects/cross-origin-objects.html` because each included test case
tests both `Location` and `WindowProxy`, the latter of which isn't
implemented correctly yet. In fact, the first test case "Basic sanity-
checking" passes if it's reduced to only check `Location` as follows:

    addTest(function(win) {
      assert_equals(B.location.pathname, path, "location.href works same-origin");
      assert_throws("SecurityError", function() { win.location.pathname; }, "location.pathname throws cross-origin");
    }, "Basic sanity-checking");

yvt committed:
feat(script): Implement `[[Set]]` for `Location`

yvt committed:
doc(script): improve comments in `proxyhandler.rs`

yvt committed:
feat(script): implement the last step of `CrossOriginOwnPropertyKeys`

yvt committed:
feat(script): implement `getOwnEnumerablePropertyKeys` for `Location`

Fixes the following assertion from `tests/wpt/web-platform-tests/html/
browsers/origin/cross-origin-objects/cross-origin-objects.html`:

    assert_equals(Object.keys(win.location).length, 0,
                        "Object.keys() gives the right answer for cross-origin Location");

yvt committed:
style(script): address `test-tidy` errors

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/codegen/CodegenRust.py, components/script/script_runtime.rs, components/script/dom/bindings/interface.rs, components/script/dom/bindings/mod.rs, components/script/dom/webidls/Location.webidl and 7 more
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/script_runtime.rs, components/script/dom/bindings/interface.rs, components/script/dom/bindings/mod.rs, components/script/dom/webidls/Location.webidl and 7 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented:
This is very exciting! I'll have a closer look at the changes when I'm more awake, but thank you for splitting them up like this!

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit d733abfca02cdb9fd2af4f0d82ff050e25f71829 with merge 2b2b5c6d8ed3eb875f5c03efadfde6018138f81d...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/runs/3092727606)

jdm commented:
A timeout in both cross-origin-objects and cross-origin-objects-in-new-window is interesting.

yvt commented:
`cross-origin-objects` passes locally but takes a considerable time (30+ seconds). Maybe CI is too slow?

`cross-origin-objects-in-new-window` is probably just #26299 - It most of the time passes but can timeout randomly.

jdm commented:
Ah yes, CI being slow enough to trigger timeouts sounds very plausible.

jdm commented on components/script/dom/bindings/principals.rs:
```diff
@@ -0,0 +1,122 @@
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(&self.0 as *const NonNull<JSPrincipals> as *const ServoJSPrincipals) }
```
Let's do `&*(self.0.as_ptr().cast::<ServoJSPrincipals>())` instead.

yvt replied:
That would change the code's semantics - it's intended to reinterpret `&self.0: &NonNull<JSPrincipals>` (not `self.0`) as `&ServoJSPrincipals` (i.e., `&NonNull<JSPrincipals>`).

yvt replied:
This approach lets us have `ServoJSPrincipalsRef: Copy`, but on second thought, this might not be worth the extra `unsafe` block. I implemented a different approach in commit e98cba1896160e6512d8f74988d253cd5b1047e4.

jdm commented on components/script/dom/bindings/interface.rs:
```diff
@@ -9,7 +9,9 @@ use crate::dom::bindings::codegen::PrototypeList;
 use crate::dom::bindings::guard::Guard;
+use crate::dom::bindings::principals::ServoJSPrincipals;
 use crate::dom::bindings::utils::{ProtoOrIfaceArray, DOM_PROTOTYPE_SLOT};
+use crate::dom::window::Window;
```
This looks unused.

jdm commented on components/script/dom/bindings/interface.rs:
```diff
@@ -36,6 +38,8 @@ use js::rust::wrappers::{JS_DefineProperty3, JS_DefineProperty4, JS_DefineProper
 use js::rust::wrappers::{JS_LinkConstructorAndPrototype, JS_NewObjectWithGivenProto};
 use js::rust::{define_methods, define_properties, get_object_class};
 use js::rust::{HandleObject, HandleValue, MutableHandleObject, RealmOptions};
+use libc;
```
This looks unused.

jdm commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5751,7 +5952,75 @@ def getBody(self):
+        CGAbstractExternMethod.__init__(self, descriptor, "set", "bool", args)
+        self.descriptor = descriptor
+
+    def getBody(self):
```
It looks like this function is the same for every proxy class that generates it. If that's the case, it could just be a common function in dom/bindings/proxyhandler.rs, rather than part of the codegen for each type.

yvt replied:
Addressed in 77012b72821fa197581789a32cda97f66b898c20

jdm commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -5751,7 +5952,75 @@ def getBody(self):
+        return CGGeneric(self.getBody())
+
+
+class CGDOMJSProxyHandler_getPrototype(CGAbstractExternMethod):
```
Same here - it looks like this could be a common function in dom/bindings/proxyhandler.rs instead.

yvt replied:
Addressed in 591ce9e4168738c4bdac0ecbd8e9155101644ba2

yvt replied:
Nevermind - `CGDOMJSProxyHandler_getPrototype` actually requires a function pointer to an interface-specific `GetProtoObject` implementation.

jdm commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -6401,22 +6673,29 @@ def reexportedName(name):
+                if descriptor.isMaybeCrossOriginObject():
+                    cgThings.append(CGDOMJSProxyHandler_getPrototype(descriptor))
+                    cgThings.append(CGDOMJSProxyHandler_set(descriptor))
+                    pass
```
nit: remove the `pass`.

yvt replied:
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

jdm reviewed (changes requested):
Fantastic work, especially all the comments!

yvt committed:
refactor(script): apply suggestion

Co-authored-by: Josh Matthews <josh@joshmatthews.net>

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+
+unsafe fn get_getter_object(d: &PropertyDescriptor, out: RawMutableHandleObject) {
+    if (d.attrs & jsapi::JSPROP_GETTER as u32) != 0 {
+        out.set(std::mem::transmute(d.getter));
```
No, that doesn't work because it's [a nullable function pointer][1]. A non-transmuting alternative is verbose: `d.getter.map(|fptr| fptr as usize as _).unwrap_or_else(ptr::null_mut)`.

(The transmutation requirement was removed in [a newer version of SpiderMonkey][2] which we don't have yet.)

[1]: https://searchfox.org/mozilla-central/rev/cfbd0c553830609bab71a6f5550fccdffa86ce99/js/public/Class.h#339-357
[2]: https://hg.mozilla.org/integration/autoland/rev/2d857a9fa2ca
[3]: https://searchfox.org/mozilla-central/rev/cfbd0c553830609bab71a6f5550fccdffa86ce99/js/public/PropertyDescriptor.h#187-197

yvt committed:
refactor(script): apply suggestions

yvt commented on components/script/dom/bindings/interface.rs:
```diff
@@ -9,7 +9,9 @@ use crate::dom::bindings::codegen::PrototypeList;
 use crate::dom::bindings::guard::Guard;
+use crate::dom::bindings::principals::ServoJSPrincipals;
 use crate::dom::bindings::utils::{ProtoOrIfaceArray, DOM_PROTOTYPE_SLOT};
+use crate::dom::window::Window;
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/interface.rs:
```diff
@@ -36,6 +38,8 @@ use js::rust::wrappers::{JS_DefineProperty3, JS_DefineProperty4, JS_DefineProper
 use js::rust::wrappers::{JS_LinkConstructorAndPrototype, JS_NewObjectWithGivenProto};
 use js::rust::{define_methods, define_properties, get_object_class};
 use js::rust::{HandleObject, HandleValue, MutableHandleObject, RealmOptions};
+use libc;
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+            jsstr.get()
+        })
+        .filter(|jsstr| !jsstr.is_null())
+        .map(|jsstr| crate::dom::bindings::conversions::jsstring_to_str(*cx, jsstr))
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+
+impl CrossOriginProperties {
+    /// Enumerate the property keys defined by `self`.
+    fn keys(&self) -> impl Iterator<Item = *const std::os::raw::c_char> + '_ {
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+    // TODO: Once we have the slot for the holder, it'd be more efficient to
+    //       use `ensure_cross_origin_property_holder`.
+    *bp = if let Some(key) =
+        crate::dom::bindings::conversions::jsid_to_string(*cx, Handle::from_raw(id))
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+    crate::dom::bindings::conversions::jsid_to_string(*cx, Handle::from_raw(id))
+        .filter(|st| st == "then")
+        .is_some() ||
+        {
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+    props: RawMutableHandleIdVector,
+) {
+    rooted!(in(*cx) let mut id: jsid);
+    {
```
Addressed in 690d8462a5105f4662c3c0bc6ab0c0d8c414b040

yvt committed:
refactor(script): don't conjure up `ServoJSPrincipals` in `ServoJSPrincipalsRef::deref`

It's technically safe to do because of `#[repr(transparent)]` and is a
prerequisite of having `ServoJSPrincipalsRef: Copy`, but I guess it's
not worth an `unsafe` block.

yvt commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -1941,12 +1982,24 @@ def template(m):
-            '    JSPropertySpec::ZERO',
-            'JSPropertySpec',
-            condition, specData)
+        if self.crossorigin:
+            groups = groupby(array, lambda m: condition(m, self.descriptor))
+            assert len(list(groups)) == 1  # can't handle mixed condition
+            elems = [template(m) % specData(m) for m in array]
+            return dedent(
+                """
+                const %s: &[JSPropertySpec] = &[
+                %s,
+                    JSPropertySpec::ZERO,
+                ];
+                """) % (name, ',\n'.join(elems))
+        else:
+            return self.generateGuardedArray(
+                array, name,
+                template,
+                '    JSPropertySpec::ZERO',
+                'JSPropertySpec',
+                condition, specData)
```
Do you mean adding a new method similar to `PropertyDefiner.generateGuardedArray`? If that's the case, `generateFlatArray` or `generateUnguardedArray` might be a better name since anything in `PropertyDefiner` obviously pertains to `JSFunctionSpec` or `JSPropertySpec`.

jdm replied:
Yeah, wherever makes sense. I just saw very similar code being added for JSPropertySpec and JSFunctionSpec arrays, and it would be nice to avoid duplicating that logic.

yvt replied:
Addressed in 2e0dd0816f8385fda04e69908916e0fe00fb6d39

yvt committed:
doc(script): fix comments

`History` is not a maybe-cross-origin object. I must have been very
sleepy when I wrote this.

yvt committed:
refactor(script): squash `CGDOMJSProxyHandler_set`

The implementation in `crate::dom::bindings::proxyhandler::
maybe_cross_origin_set_rawcx` is now directly assigned to `ProxyTraps::
set`.

yvt committed:
refactor(script): refactor common code into `PropertyDefiner.generateUnguardedArray`

There are code fragments in `(Method|Attr)Definer.generateArray` that
are much alike. This commit refactors them into a new method of
`PropertyDefiner` named `generateUnguardedArray` (in contrast to the
existing method `generateGuardedArray`).

yvt committed:
refactor(script): make grouping clearer while keeping `test-tidy` happy

yvt commented on components/script/dom/bindings/codegen/CodegenRust.py:
```diff
@@ -1874,8 +1913,10 @@ def getter(attr):
+
+            if ((attr.readonly and not attr.getExtendedAttribute("PutForwards")
+                    and not attr.getExtendedAttribute("Replaceable"))
+                    or (self.crossorigin and not attr.getExtendedAttribute("CrossOriginReadable"))):
```
Made some attempt in b6ee398b91a4a956905481d25dffbb6b7fecdf27.

yvt committed:
style(script): add underscore to unused parameter

yvt commented on components/script/dom/windowproxy.rs:
```diff
@@ -1084,6 +1084,9 @@ pub fn new_window_proxy_handler() -> WindowProxyHandler {
 // These traps often throw security errors, and only pass on calls to methods
 // defined in the DissimilarOriginWindow IDL.
 
+// TODO: reuse the infrastructure in `proxyhandler.rs`. For starters, the calls
```
Opened #28556.

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+    cross_origin_properties: &'static CrossOriginProperties,
+    out_holder: RawMutableHandleObject,
+) -> bool {
+    // TODO: We don't have the slot to store the holder yet. For now,
```
Opened #28557.

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+    id: RawHandleId,
+    bp: *mut bool,
+) -> bool {
+    let _ = proxy;
```
Addressed in 110b3ab6bcd8a25d56fcce3513b2c14072532598.

yvt committed:
fix(script): don't pass an unrooted slice to `from_rooted_slice`

yvt commented on components/script/dom/bindings/proxyhandler.rs:
```diff
@@ -177,3 +199,535 @@ pub fn fill_property_descriptor(
+        *cx,
+        receiver,
+        setter_jsval.handle().into(),
+        &jsapi::HandleValueArray::from_rooted_slice(&[v.get()]),
```
Addressed in 42574ea74d63e80694062dc63512da13d3d2b05e.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit afbe2fa has been approved by `jdm`

<!-- @bors-servo r=jdm afbe2fa1f259411b6c49a3ec3b2baccfbf1664d9 -->

bors-servo commented:
:hourglass: Testing commit afbe2fa1f259411b6c49a3ec3b2baccfbf1664d9 with merge bd92fad81a24d08208a5739cad4bde6eb58d6ce8...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/runs/3214631305), [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/HTK6VwoGSf-ECVoL-4D9FQ)
Approved by: jdm
Pushing bd92fad81a24d08208a5739cad4bde6eb58d6ce8 to master...
